File: c/a2024/22.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <limits.h>

static inline int step(int n) {
    n = ((n << 6) ^ n) % 16777216;
    n = ((n >> 5) ^ n) % 16777216;
    n = (((n << 11) & INT_MAX) ^ n) % 16777216;
    return n;
}

int main() {
    long ans1 = 0;

    FILE* inp = fopen("../../inputs/a2024/22.txt", "r");

    int d[130321] = {0};
    int seen_now[130321];

    char buf[20];

    while (fscanf(inp, "%s\n",  buf) == 1) {
        int n = atoi(buf);
        int window = 0;
        for (int i=0; i<4; ++i) {
            int n_ = step(n);
            window = window*19 + (n_ % 10) - (n % 10) + 9;
            n = n_;
        }

        memset(seen_now, 0, sizeof(seen_now));

        for (int i=0; i<1997; ++i) {
            if (i == 1996) {
                ans1 += (long) n;
            }
            if (seen_now[window] == 0) {
                d[window] += n % 10;
                seen_now[window] = 1;
            }

            int n_ = step(n);
            window = (window*19 + (n_ % 10) - (n % 10) + 9) % 130321;
            n = n_;
        }
    }

    int ans2 = 0;
    for (int i=0; i<130321; ++i) {
        if (d[i] > ans2) {
            ans2 = d[i];
        }
    }
    printf("part 1: %ld\n", ans1);
    printf("part 2: %i\n", ans2);
}

File: python/a2024/22.py
from problem import get_problem, get_problem_lines, look
from collections import defaultdict
from time import time
#aaa = time()

def step(n):
    n = ((n * 64) ^ n) % 16777216
    n = ((n // 32) ^ n) % 16777216
    n = ((n * 2048) ^ n) % 16777216
    return n

ans1 = 0
d = defaultdict(int)

for line in get_problem_lines():
    n = int(line)
    window = 0
    for _ in range(4):
        n_ = step(n)
        window = window*20 + ((n_ % 10) - (n % 10) + 10)
        n = n_

    seen_now = set()
    for i in range(1997):
        if i == 1996:
            ans1 += n

        if window not in seen_now:
            d[window] += n % 10
            seen_now.add(window)

        n_ = step(n)
        window = (window*20 + ((n_ % 10) - (n % 10) + 10)) % (20 ** 4)
        n = n_

#print(time() - aaa)
print(f"part 1: {ans1}")
print(f"part 2: {max(d.values())}")
